### PR TITLE
Add Safe Abstractions for `FuriStreamBuffer`

### PR DESCRIPTION
**Pull request by cptpiepmatz**
Hey 👋🏻,
In this PR, I added some safe abstractions for the `FuriStreamBuffer`.

While working on it, I came across this note in the stream buffer implementation:
```
 * ***NOTE***: Stream buffer implementation assumes there is only one task or
 * interrupt that will write to the buffer (the writer), and only one task or
 * interrupt that will read from the buffer (the reader).
```

So, I figured, why not enforce this constraint in the type system? The result is a split into a sender and receiver, similar to the `mpsc` channel in the standard library. They implement `Send`, so they can be moved between threads, but not `Clone` or `Sync`, making sure we only have one of each in a pair. Unfortunately, I had to use `Arc` to give ownership to both sides, which means this requires `alloc`. Other than that, I’m happy with how the API turned out.

I also added an example where data is sent between different threads.

### CODE REVIEW

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (JarvisCraft) [inline crates/flipperzero/src/furi/stream_buffer.rs:143]: Safety comment would be useful here

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (JarvisCraft) [inline crates/flipperzero/src/furi/stream_buffer.rs:271]: Same here

**Review by JarvisCraft — CHANGES_REQUESTED** [review]
The API looks great. Here is the first iteration of nitpicking, but generally the PR looks super accurate

**Comment by JarvisCraft** [comment]
It will take me some time to think on the `Arc`-related part, but it looks good for the start

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs:38]: Did that, the `NonNull` is handled by using the raw pointer for the field.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs:276]: Did that, the `NonNull` is handled by using the raw pointer for the field.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: I replaced them with local variables.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: I dislike it visually with the if-else but I'm ok with it.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: Removed that one.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: I just tried it out, it doesn't seem that the flipper checks if the `trigger_level` is stupid in the alloc. 🤷🏻

File: crates/flipperzero/src/furi/mod.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/mod.rs]: It feels really bloated to a feature just because of the `Arc` I used. I could replace it with some custom counting logic using an `AtomicUsize` and `sys::malloc` to get a pointer but I'm not sure if that is better.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: I moved them all into local variables.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs:135]: I did that.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs:158]: I don't know what to really say here. Usually you would argue why `Send` is sound at this point but that's not really the reason that this type implements `Send`. Maybe I need to write quite some paragraph about the `Send` here, just like in `StreamBuffer`.

Not sure.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (dcoles) [inline crates/flipperzero/src/furi/stream_buffer.rs:429]: Perhaps `stream` might be a better name, since you're not actually getting a `StreamBuffer`—you're getting an abstraction on top of it.

**Review by dcoles — CHANGES_REQUESTED** [review]
Hi @cptpiepmatz!

Thanks for the PR! I really like the `Sender` / `Receiver` concept and how extensively you've documented everything 💖.

I think it would be useful to expose the low-level `StreamBuffer` API for `no-alloc` cases, but that's something we can certainly expose after merging this PR. Vast majority of comments are just little things for consistency with the other parts of the `flipperzero-rs` API.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (dcoles) [inline crates/flipperzero/src/furi/stream_buffer.rs:12]: It would probably be good to have this in `furi::time`. There is a `MAX_DURATION_TICKS` (`0x7FFFFFFFF`) though that has a slightly different purpose.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: I replaced that.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs:78]: Added that safety comment.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: Yep, removed the specific type.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: Is a `furi::Result` now, I also removed the `EmtpyResult` as it became obsolete.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: That is true, I removed that one.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: I handled this by adding a `as_stream_buffer` and `into_stream_buffer` methods on the `Sender` and `Receiver`.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: This is replaced with `StreamBuffer::into_stream` now.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: I added `WAIT_FOREVER` as a new constant.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (cptpiepmatz) [reply crates/flipperzero/src/furi/stream_buffer.rs]: I made `StreamBuffer` `Send` and `Sync` with some comments why that is safe. For that I marked `send` and `receive` of the `StreamBuffer´  as `unsafe`.

**Comment by cptpiepmatz** [comment]
I made the `StreamBuffer` public with all of its methods, the `send` and `receive` methods are `unsafe` as they need to uphold the guarantee that only writer and one reader can exists at a time.

The `Sender` and `Receiver` is moved in the private `stream` module. I re-exported the entire module but that allowed to just mark it as `#[cfg(feature = "alloc")]`. For that purpose I also removed the `stream-buffer´ feature again. I like this setup now, you can use the `StreamBuffer` without `alloc` but if you don't want to manage `send` and `receive` manually, you can use `Sender` and `Receiver` by enabling the `alloc` feature.

Since the `StreamBuffer` itself is now public, I added some methods on the `Sender` and `Receiver` to access the `StreamBuffer` and removed all the redirect methods. I also heavily updated the documentation to incorporate the new API.

**Review by dcoles — APPROVED** [review]

**Comment by dcoles** [comment]
Hi @cptpiepmatz. This looks great! I'd be happy to merge this once @JarvisCraft has taken a final look.

**Comment by cptpiepmatz** [comment]
@dcoles could you please approve the workflow again? I added a `SetTriggerLevelError` to make clippy happy

**Comment by cptpiepmatz** [comment]
@JarvisCraft is this ready to merge?

**Comment by JarvisCraft** [comment]
> @JarvisCraft is this ready to merge?

Hi there, apologies for the long reply. I will take a look at the PR this week.

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (JarvisCraft) [inline crates/flipperzero/src/furi/stream_buffer.rs:74]: Non-blocker nit (ptobably, for the future changes): given that type annotation is already used, `from` seems more appropriate:

```suggestion
        let size = usize::from(size);
```

File: crates/flipperzero/src/furi/stream_buffer.rs
Comment (JarvisCraft) [inline crates/flipperzero/src/furi/stream_buffer.rs:245]: For future work on our side: this should probably be moved to some internal `utils` module

**Review by JarvisCraft — APPROVED** [review]
Apologies for the delay on code-review.

The PR looks amazing and the detailed docs are just the chef's kiss.

Thanks a lot for your contribution <3

**Comment by JarvisCraft** [comment]
Squashed the commits so that `fix`-named  intermediate comments don't end up in `main`

**Pull request merged**
